samples/ask-akka-agent/src/main/java/akka/ask/indexer/application/RagIndexingWorkflow.java

```diff
@@ -143,33 +142,30 @@ public Effect<Done> abort() {
   }
 
   // tag::def[]
+
+
   @Override
-  public WorkflowDef<State> definition() {
-    return workflow()
-        .addStep(processingFileStep())
-        // the processing step might take a while
-        .defaultStepTimeout(Duration.of(1, MINUTES));
+  public WorkflowConfig configuration() {
+    return WorkflowConfig.builder()
+      .defaultStepTimeout(ofMinutes(1))
+      .build();
   }
 
-  private Step processingFileStep() {
-    return step(PROCESSING_FILE_STEP) // <1>
-        .call(() -> {
-          if (currentState().hasFilesToProcess()) {
-            indexFile(currentState().head().get());
-          }
-        })
-        .andThen(() -> {
-          // we need to check if it hasFilesToProcess, before moving the head
-          // because if workflow is aborted, the state is cleared, and we won't have
-          // anything in the list
-          if (currentState().hasFilesToProcess()) { // <2>
-            var newState = currentState().headProcessed();
-            logger.debug("Processed {}/{}", newState.totalProcessed(), newState.totalFiles());
-            return effects().updateState(newState).transitionTo(PROCESSING_FILE_STEP); // <3>
-          } else {
-            return effects().pause(); // <4>
-          }
-        });
+  @StepName("processing-file")
+  private StepEffect processingFileStep() { // <1>
+    if (currentState().hasFilesToProcess()) {
+      indexFile(currentState().head().get());
+    }
+
+    if (currentState().hasFilesToProcess()) { // <2>
+      var newState = currentState().headProcessed();
+      logger.debug("Processed {}/{}", newState.totalProcessed(), newState.totalFiles());
+      return stepEffects()
+        .updateState(newState)
+        .thenTransitionTo(RagIndexingWorkflow::processingFileStep); // <3>
+    } else {
+      return stepEffects().thenPause(); // <4>
+    }
```

Comment:
We can later improve this logic by calling processingFileStep with a list of files to process. 

Pick the head to index and transition to itself using only the list tail. Until we reach an empty list.